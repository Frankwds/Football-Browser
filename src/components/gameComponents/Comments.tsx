import { ChevronRightIcon } from "@chakra-ui/icons";

import {
  InputGroup,
  InputRightElement,
  Button,
  Input,
  Flex,
  Text,
} from "@chakra-ui/react";
import { useQuery, useMutation, gql } from "@apollo/client";
import { useState } from "react";

export const COMMENTONGAME = gql`
  mutation Mutation($gameId: String!, $comment: String!) {
    commentOnGame(gameID: $gameId, comment: $comment) {
      id_odsp
      comments
    }
  }
`;

type Props = {
  id: string;
  comments: string[];
  callbackFunc: (comments: string[]) => void;
};

/**
 * Comments - Presents activity feed
 *
 * @param isAdmin
 *
 * @returns {React.FC}
 *
 * Enables users to see and add new activities. Extends the properties of admin-users to also be able to edit and delete activities.
 * Gives possibilites to sort activities on date, maxParticipants and lexicographic order.
 *
 * Handles persistence and retrieves data through ../api/firebase.ts
 */

const Comments: React.FC<Props> = (props) => {
  const [comments, setComments] = useState(props.comments);

  const [messageInput, setMessageInput] = useState("");

  const externalMutation = (data: any) => {
    props.callbackFunc(data.data.commentOnGame.comments);
  };

  const [mutateFunction, { data, loading, error }] = useMutation(
    COMMENTONGAME,
    {
      onCompleted: (data) => {
        externalMutation({ data });
      },
    }
  );
  if (loading) return <p>Loading...</p>;
  if (error) return <p>StupidFrankError :(</p>;

  const handleSendMessage = () => {
    console.log("Yo sendin a message");
    if (messageInput === "") return;
    setComments([...comments, messageInput]);

    mutateFunction({ variables: { gameId: props.id, comment: messageInput } });

    setMessageInput("");
  };

  return (
    <div id="commentComponent" key={comments.length}>
      <InputGroup size="md">
        <Input
          testid="comment-input-field"
          border="1px solid black"
          marginRight="5px"
          placeholder="Skriv en melding"
          //value={messageInput}
          onChange={(e) => setMessageInput(e.target.value)}
        ></Input>
        <InputRightElement width="6.5rem">
          <Button
          testid = "commentButton"
            backgroundColor="teal"
            h="2rem"
            size="sm"
            onClick={handleSendMessage}
            rightIcon={<ChevronRightIcon fontSize="24px" />}
          >
            Comment
          </Button>
        </InputRightElement>
      </InputGroup>
      {comments.map((comment, index) => {
        return (
          <Flex key={comment+index}>
            <Text padding="10px" borderRadius="10px" fontWeight="bold">
              {comment}
            </Text>
          </Flex>
        );
      })}
    </div>
  );
};
export default Comments;
