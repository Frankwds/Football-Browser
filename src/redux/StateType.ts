import {SearchData} from "./searchFilter/searchFilterTypes"; 

//A type interface for the only implemented query, scalable with more reducers.
export interface StateType{
    searchQueries: SearchData,
}


