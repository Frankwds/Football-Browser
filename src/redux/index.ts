import { combineReducers } from "redux";
import {searchReducer} from "./searchFilter/filterSearchReducer";


// In this case only one reducer, but very scalable.
const allReducers = combineReducers({
  searchQueries: searchReducer,
});

export default allReducers;