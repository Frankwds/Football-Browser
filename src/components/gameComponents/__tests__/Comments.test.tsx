import ReactDOM from "react-dom";
import { act } from "react-dom/test-utils";
import Comments from "../Comments";
import renderer from "react-test-renderer";
import { createStore } from "redux";
import allReducers from "../../../redux";
import { Provider } from "react-redux";
import { ChakraProvider } from "@chakra-ui/react";

// Test <div> container
let container: any;

// Set up redux store
const store = createStore(allReducers);

beforeEach(() => {
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  document.body.removeChild(container);
  container = null;
});

describe("Testing Comments", () => {
  it("should render without crashing", () => {
    act(() => {
      ReactDOM.render(
        <Provider store={store}>
          <ChakraProvider>
            <Comments
              id={"testId"}
              comments={["hei", "og", "hopp"]}
              callbackFunc={() => {}}
            />
          </ChakraProvider>
        </Provider>,
        container
      );
    });
  });

  it("snapshot should be same as previous", () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <ChakraProvider>
            <Comments
              id={"testId"}
              comments={["hei", "og", "hopp"]}
              callbackFunc={() => {}}
            />
          </ChakraProvider>
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
