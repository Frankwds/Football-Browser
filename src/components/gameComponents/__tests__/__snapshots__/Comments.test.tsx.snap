// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Comments snapshot should be same as previous 1`] = `
Array [
  <div
    id="commentComponent"
  >
    <div
      className="chakra-input__group css-4302v8"
    >
      <input
        className="chakra-input css-cv4qwk"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Skriv en melding"
        testid="comment-input-field"
      />
      <div
        className="chakra-input__right-element css-dzvope"
        size="md"
      >
        <button
          className="chakra-button css-8wtj1w"
          onClick={[Function]}
          testid="commentButton"
          type="button"
        >
          Comment
          <span
            className="chakra-button__icon css-1hzyiq5"
          >
            <svg
              aria-hidden={true}
              className="chakra-icon css-13ewnpa"
              focusable={false}
              viewBox="0 0 24 24"
            >
              <path
                d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                fill="currentColor"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
    <div
      className="css-k008qs"
    >
      <p
        className="chakra-text css-12m61xp"
      >
        hei
      </p>
    </div>
    <div
      className="css-k008qs"
    >
      <p
        className="chakra-text css-12m61xp"
      >
        og
      </p>
    </div>
    <div
      className="css-k008qs"
    >
      <p
        className="chakra-text css-12m61xp"
      >
        hopp
      </p>
    </div>
  </div>,
  <span />,
]
`;
